### PR TITLE
Removing/Updating old publisher, ra, and pubsub resources

### PR DESCRIPTION
**Pull request by nachocano**
Fixes https://github.com/google/knative-gcp/issues/1217

## Proposed Changes

- 🧽  Removing old PS and Topics k8s objects and re-create them. Detecting whether it was old or not based on changes to spec.topic between the desired and the one stored. The only way this can happen is when upgrading to 0.16, as spec.topic is immutable. Deleting the wrapper objects, will take care of deleting their owned objects (RA, publisher's, scaled objects).
- 🧽  Updating scheduler target pubsub topic.
- 🧽  Deleting and re-creating the storage notification if appropriate. Cannot find a way to update the existing notification

I first tried deleting the PubSub resources instead of our wrappers K8s objects, but that has the problem that we cannot update them to use the newer topic names, as spec.topic is immutable.

**Release Note**

```release-note
Given that we have renamed many resources in https://github.com/google/knative-gcp/pull/1207, when upgrading to 0.16 we will delete those old resources and re-create them with the newer names. You may experience some delay in getting your resources back to the ready state.
```

**Docs**

### CODE REVIEW

**Comment by nachocano** [comment]
This won't actually work as the topic.spec.topic is immutable...

**Comment by nachocano** [comment]
/assign @Harwayne 
so that you can start to take a look. I haven't fully tested it, will create a new cluster with 0.15, create some resources there, and then apply these changes. 

Integration are still failing due to monitoring, but that should have been fixed. Need to wait for 1 hour or so until we get a new nightly

File: pkg/reconciler/events/auditlogs/auditlogs.go [hunk 1]
```diff
@@ -92,6 +92,12 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, s *v1beta1.CloudAuditLog
 	}
 	s.Status.StackdriverSink = sink
 	s.Status.MarkSinkReady()
+
+	// TODO remove after 0.16 cut.
+	if err := c.deleteOldSink(ctx, s); err != nil {
+		return reconciler.NewEvent(corev1.EventTypeWarning, "DeleteSinkFailed", "Failed to delete StackDriver sink: %s", err.Error())
```
Comment (Harwayne) [inline pkg/reconciler/events/auditlogs/auditlogs.go:98]:
```suggestion
		return reconciler.NewEvent(corev1.EventTypeWarning, "DeleteOldSinkFailed", "Failed to delete old StackDriver sink: %s", err.Error())
```

File: pkg/reconciler/events/auditlogs/auditlogs.go [hunk 2]
```diff
@@ -204,3 +210,18 @@ func (c *Reconciler) FinalizeKind(ctx context.Context, s *v1beta1.CloudAuditLogs
...
+// TODO remove after 0.16 cut.
+func (c *Reconciler) deleteOldSink(ctx context.Context, s *v1beta1.CloudAuditLogsSource) error {
+	logadminClient, err := c.logadminClientProvider(ctx, s.Status.ProjectID)
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Failed to create LogAdmin client", zap.Error(err))
+		return err
+	}
+	oldSinkName := fmt.Sprintf("sink-%s", string(s.UID))
+	if err = logadminClient.DeleteSink(ctx, oldSinkName); status.Code(err) != codes.NotFound {
+		logging.FromContext(ctx).Desugar().Error("Failed to delete StackDriver sink", zap.String("sinkName", oldSinkName), zap.Error(err))
```
Comment (Harwayne) [inline pkg/reconciler/events/auditlogs/auditlogs.go:223]:
```suggestion
		logging.FromContext(ctx).Desugar().Error("Failed to delete old StackDriver sink", zap.String("sinkName", oldSinkName), zap.Error(err))
```

File: pkg/reconciler/messaging/channel/channel.go [hunk 3]
```diff
@@ -286,22 +300,12 @@ func (r *Reconciler) syncSubscribersStatus(ctx context.Context, channel *v1beta1
...
-	}
-	if topic != nil {
-		if topic.Status.Address != nil {
-			channel.Status.SetAddress(topic.Status.Address.URL)
-		} else {
-			channel.Status.SetAddress(nil)
-		}
-		return topic, nil
-	}
+
```
Comment (Harwayne) [inline pkg/reconciler/messaging/channel/channel.go:303]: It feels odd for this to be separated from the `if err != nil` line, can it be moved just before that one?

File: pkg/reconciler/messaging/channel/channel.go [hunk 4]
```diff
@@ -310,14 +314,60 @@ func (r *Reconciler) reconcileTopic(ctx context.Context, channel *v1beta1.Channe
 		Annotations:        resources.GetTopicAnnotations(clusterName),
 	})
 
-	topic, err = r.RunClientSet.InternalV1beta1().Topics(channel.Namespace).Create(t)
-	if err != nil {
-		logging.FromContext(ctx).Desugar().Error("Failed to create Topic", zap.Error(err))
-		r.Recorder.Eventf(channel, corev1.EventTypeWarning, "TopicCreateFailed", "Failed to created Topic %q: %s", topic.Name, err.Error())
-		return nil, err
+	if apierrs.IsNotFound(err) {
```
Comment (Harwayne) [inline pkg/reconciler/messaging/channel/channel.go:317]: Isn't this still the error for not getting the Topic?

File: pkg/reconciler/messaging/channel/channel.go [hunk 5]
```diff
@@ -326,10 +376,6 @@ func (r *Reconciler) getTopic(_ context.Context, channel *v1beta1.Channel) (*int
 	if err != nil {
 		return nil, err
 	}
-	if !metav1.IsControlledBy(topic, channel) {
```
Comment (Harwayne) [inline pkg/reconciler/messaging/channel/channel.go:329]: Why is this being removed?

File: pkg/reconciler/events/auditlogs/auditlogs.go (hunk 1 above)
Comment (nachocano) [reply pkg/reconciler/events/auditlogs/auditlogs.go:98]: Done, thanks!

File: pkg/reconciler/events/auditlogs/auditlogs.go (hunk 2 above)
Comment (nachocano) [reply pkg/reconciler/events/auditlogs/auditlogs.go:223]: Done, thanks!

File: pkg/reconciler/messaging/channel/channel.go (hunk 3 above)
Comment (nachocano) [reply pkg/reconciler/messaging/channel/channel.go:303]: which one do you mean? the `if topic != nil` stuff?

File: pkg/reconciler/messaging/channel/channel.go (hunk 4 above)
Comment (nachocano) [reply pkg/reconciler/messaging/channel/channel.go:317]: I refactored this a bit to make it more consistent with other reconcilers. In general, I think we need to do a big refactor to this old channel code. Is pretty hard to follow...
If the error is notFound, then we create the Topic... 
Previously we were doing something like: `if err != nil && !apierrors.IsNotFound(err)`...  now I'm checking for the positive case of notFound.
Does it make sense?

File: pkg/reconciler/messaging/channel/channel.go (hunk 5 above)
Comment (nachocano) [reply pkg/reconciler/messaging/channel/channel.go:329]: Moved it to the caller code to better match other reconcilers. Do you prefer it here?

File: pkg/reconciler/messaging/channel/channel.go (hunk 4 above)
Comment (Harwayne) [reply pkg/reconciler/messaging/channel/channel.go:317]: Yes.

File: pkg/reconciler/messaging/channel/channel.go (hunk 5 above)
Comment (Harwayne) [reply pkg/reconciler/messaging/channel/channel.go:329]: This is fine, thanks for explaining.

File: pkg/reconciler/messaging/channel/channel.go (hunk 3 above)
Comment (Harwayne) [reply pkg/reconciler/messaging/channel/channel.go:303]: I mean that we set `err` here on line 302, but don't check it until line 317.

Comment (nachocano) [reply pkg/reconciler/messaging/channel/channel.go:303]: got it, indeed, it does make sense. Moved it

**Comment by nachocano** [comment]
Synced with @grac3gao, it seems that the previous run picked a cluster with no workload identity enabled, so all tests failed... 
re-running

**Comment by nachocano** [comment]
Installed 0.15, then created a few sources and a channel... Sent events, all worked...
Then did a ko apply from my branch... Things were removed properly and newer stuff was created. Examples:

CAL

topic: cloudauditlogssource-d56c6c3d-4ab4-4f38-8ded-9b207f51f1c5  --> cre-src_default_test-auditlogs_d56c6c3d-4ab4-4f38-8ded-9b207f51f1c5	
sub: cre-pull-0b062770-0ed5-4e14-bd81-e0948661ef96  --> cre-src_default_test-auditlogs_11819cd7-6b73-4bb8-bb72-e5e838563a45
ksvc: cre-test-auditlogs-publish  --> removed
ra: cre-pull-0b062770-0ed5-4e14-bd81-e0948661ef96-67c8ff6d57-9798q --> cre-src-test-auditlogs-11819cd7-6b73-4bb8-bb72-e5e838563a45llns

--------

scheduler

topic: scheduler-0867c821-cf1c-4c2e-9909-5087f0d9e87e --> cre-src_default_test-scheduler_0867c821-cf1c-4c2e-9909-5087f0d9e87e
sub: cre-pull-ea0d5c66-9572-442f-a63e-33c96dc09fc5  --> cre-src_default_test-scheduler_34ad9c55-2498-4120-892e-8be2183d3c39
ksvc: cre-test-scheduler-publish  --> removed 
ra: cre-pull-ea0d5c66-9572-442f-a63e-33c96dc09fc5-67fcf95c79-h5mx2  --> cre-src-test-scheduler-34ad9c55-2498-4120-892e-8be2183d3c387vfx

---------

storage

topic: storage-111f1ef4-b734-4335-be58-fdf083e19e8d  --> cre-src_default_test-storage_111f1ef4-b734-4335-be58-fdf083e19e8d
sub: cre-pull-03d4ce13-3d09-4ba5-90a3-c4d84d0ec7a2 --> cre-src_default_test-storage_1be82382-4225-4dc6-bede-0b0acd5a2742
ksvc: cre-test-storage-publish  -> removed
ra: cre-pull-03d4ce13-3d09-4ba5-90a3-c4d84d0ec7a2-b966cdffb-64vvk  --> cre-src-test-storage-1be82382-4225-4dc6-bede-0b0acd5a2742-kqdfh

**Comment by nachocano** [comment]
Few problems I faced. 

The controller seems to be trying to export metrics to stackdriver, even custom metrics is disabled. Not sure if this is a problem from the upgrade itself or if it's in master.
Log from the controller
```2020/07/06 22:04:03 Failed to export to Stackdriver: [rpc error: code = PermissionDenied desc = Permission monitoring.metricDescriptors.create denied (or the resource may not exist).; rpc error: code = PermissionDenied desc = Permission monitoring.timeSeries.create denied (or the resource may not exist).]```

Failing to read the auth defaults from the new configMap:
```{"level":"error","ts":"2020-07-06T23:10:10.202Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-intevents-pullsubscription-static.Reconciler","caller":"v1beta1/pubsub_defaults.go:33","msg":"Failed to get the GCPAuthDefaults","commit":"167157b","knative.dev/traceid":"cc189e25-c378-405b-b0cb-e9d33b1ac2fa","knative.dev/key":"default/test-scheduler","stacktrace":"github.com/google/knative-gcp/pkg/apis/duck/v1beta1.(*PubSubSpec).SetPubSubDefaults\n\tgithub.com/google/knative-gcp/pkg/apis/duck/v1beta1/pubsub_defaults.go:33\ngithub.com/google/knative-gcp/pkg/apis/intevents/v1beta1.(*PullSubscriptionSpec).SetDefaults\n\tgithub.com/google/knative-gcp/pkg/apis/intevents/v1beta1/pullsubscription_defaults.go:51\ngithub.com/google/knative-gcp/pkg/apis/intevents/v1beta1.(*PullSubscription).SetDefaults\n\tgithub.com/google/knative-gcp/pkg/apis/intevents/v1beta1/pullsubscription_defaults.go:36\nknative.dev/pkg/reconciler.PreProcessReconcile\n\tknative.dev/pkg@v0.0.0-20200626022628-f1ee372577e1/reconciler/reconcile_common.go:42\ngithub.com/google/knative-gcp/pkg/client/injection/reconciler/intevents/```

Failing to delete the old SD sink:
```
{"level":"error","ts":"2020-07-06T23:10:11.548Z","logger":"controller.cloud-run-events-cloudauditlogssource-controller","caller":"auditlogs/auditlogs.go:223","msg":"Failed to delete StackDriver sink","commit":"167157b","events.cloud.google.com/controller":"cloud-run-events-cloudauditlogssource-controller","auditlogsource":{"metadata":{"name":"test-auditlogs","namespace":"default","selfLink":"/apis/events.cloud.google.com/v1beta1/namespaces/default/cloudauditlogssources/test-auditlogs","uid":"d56c6c3d-4ab4-4f38-8ded-9b207f51f1c5","resourceVersion":"27216","generation":1,"creationTimestamp":"2020-07-06T21:25:26Z","annotations":{"cluster-name":"kn-upgrade","kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"events.cloud.google.com/v1beta1\",\"kind\":\"CloudAuditLogsSource\",\"metadata\":{\"annotations\":{},\"name\":\"test-auditlogs\",\"namespace\":\"default\"},
```

The Channel didn't work until I delete the subscription and created it again. 
PubSub worked as is. 
I was having a problem with Storage because I had it created with the old event type. 

Will keep on investigating

**Comment by nachocano** [comment]
Still getting errors trying to delete the SD sink. Given that is not a topic/subscription, but rather some SD internal stuff, I think is fine to leave it as is. As the code is structured, given that the sink should be in the status of the resource, it will keep on using that old sink name and things will keep on working.
For newer resources, the sink will have a more readable name.

**Comment by nachocano** [comment]
All sources working. Channel still does not mark the subscription ready when upgrading from 0.15. 
I'd suggest we get this in asap, and I keep on investigating. Need to fix the auto-update deps as well, because it's not working

**Comment by nachocano** [comment]
/hold cancel

Cancelling the hold as I was this in before the CI breaks again and I need to do a bunch of re-tests. 
The upgrade of Channel is still not working. Will try to fix it during the day (probably after fixing update deps). And if someone has cycles and can take a look at it, it'd be great. 
The other bits were working in my tests.
